File: src/test/resources/io/viash/helpers/languages/scala/test_ViashParseYaml.scala
#!/usr/bin/env scala

//> using file ../../../../../../../main/resources/io/viash/languages/scala/ViashParseYaml.scala

/*
 * Unit test for ViashYamlParser.parseYaml function
 * This test verifies that the Scala YAML parser correctly parses YAML content
 * and returns appropriate Scala data structures.
 */

// Colors for test output
object Colors {
  val RED = "\u001b[0;31m"
  val GREEN = "\u001b[0;32m"
  val YELLOW = "\u001b[1;33m"
  val NC = "\u001b[0m" // No Color
}

// Test counters
var testsPass = 0
var testsFail = 0

def runTest(testName: String, expected: Any, actual: Any): Unit = {
  print(s"Testing $testName... ")
  
  val isEqual = (expected, actual) match {
    case (None, None) => true
    case (Some(e), Some(a)) => e == a
    case (e, a) => e == a
  }
  
  if (isEqual) {
    println(s"${Colors.GREEN}PASS${Colors.NC}")
    testsPass += 1
  } else {
    println(s"${Colors.RED}FAIL${Colors.NC}")
    println(s"  Expected: $expected")
    println(s"  Actual:   $actual")
    testsFail += 1
  }
}

def testArray(testName: String, expectedArray: List[Any], actualArray: List[Any]): Unit = {
  print(s"Testing $testName... ")
  
  val isEqual = expectedArray == actualArray
  
  if (isEqual) {
    println(s"${Colors.GREEN}PASS${Colors.NC}")
    testsPass += 1
  } else {
    println(s"${Colors.RED}FAIL${Colors.NC}")
    println(s"  Expected array: [${expectedArray.mkString(", ")}]")
    println(s"  Actual array:   [${actualArray.mkString(", ")}]")
    testsFail += 1
  }
}

@main def main(): Unit = {
  println("Running ViashYamlParser Scala unit tests...")
  println()
  
  // Test 1: Basic key-value pairs
  println("=== Test 1: Basic key-value pairs ===")
  val yamlContent1 = """
par:
  input: test.txt
  number: 42
  flag: true
  empty_value:
meta:
  name: test_script
  version: 1.0
"""
  
  val result1 = ViashYamlParser.parseYaml(Some(yamlContent1))
  val par1 = result1.get("par").map(_.asInstanceOf[Map[String, Any]]).getOrElse(Map.empty)
  val meta1 = result1.get("meta").map(_.asInstanceOf[Map[String, Any]]).getOrElse(Map.empty)
  
  runTest("par.input", "test.txt", par1.getOrElse("input", None))
  runTest("par.number", 42, par1.getOrElse("number", None))
  runTest("par.flag", true, par1.getOrElse("flag", None))
  runTest("par.empty_value", None, par1.getOrElse("empty_value", None))
  runTest("meta.name", "test_script", meta1.getOrElse("name", None))
  runTest("meta.version", 1.0, meta1.getOrElse("version", None))
  println()

  // Test 2: Arrays
  println("=== Test 2: Arrays ===")
  val yamlContent2 = """
par:
  files:
    - file1.txt
    - file2.txt
    - file3.txt
  numbers:
    - 1
    - 2
    - 3
"""
  
  val result2 = ViashYamlParser.parseYaml(Some(yamlContent2))
  val par2 = result2.get("par").map(_.asInstanceOf[Map[String, Any]]).getOrElse(Map.empty)
  
  testArray("par.files", List("file1.txt", "file2.txt", "file3.txt"), par2.get("files").map(_.asInstanceOf[List[Any]]).getOrElse(List.empty))
  testArray("par.numbers", List(1, 2, 3), par2.get("numbers").map(_.asInstanceOf[List[Any]]).getOrElse(List.empty))
  println()

  // Test 3: Quoted strings
  println("=== Test 3: Quoted strings ===")
  val yamlContent3 = """
par:
  quoted_string: "This is a quoted string"
  single_quoted: 'Single quoted string'
  with_newline: "Line 1\nLine 2"
"""
  
  val result3 = ViashYamlParser.parseYaml(Some(yamlContent3))
  val par3 = result3.get("par").map(_.asInstanceOf[Map[String, Any]]).getOrElse(Map.empty)
  
  runTest("par.quoted_string", "This is a quoted string", par3.getOrElse("quoted_string", None))
  runTest("par.single_quoted", "Single quoted string", par3.getOrElse("single_quoted", None))
  runTest("par.with_newline", "Line 1\nLine 2", par3.getOrElse("with_newline", None))
  println()

  // Test 4: Boolean and numeric types
  println("=== Test 4: Boolean and numeric types ===")
  val yamlContent4 = """
par:
  bool_true: true
  bool_false: false
  integer: 123
  negative: -45
  float_val: 3.14
  negative_float: -2.71
"""
  
  val result4 = ViashYamlParser.parseYaml(Some(yamlContent4))
  val par4 = result4.get("par").map(_.asInstanceOf[Map[String, Any]]).getOrElse(Map.empty)
  
  runTest("par.bool_true", true, par4.getOrElse("bool_true", None))
  runTest("par.bool_false", false, par4.getOrElse("bool_false", None))
  runTest("par.integer", 123, par4.getOrElse("integer", None))
  runTest("par.negative", -45, par4.getOrElse("negative", None))
  runTest("par.float_val", 3.14, par4.getOrElse("float_val", None))
  runTest("par.negative_float", -2.71, par4.getOrElse("negative_float", None))
  println()

  // Test 5: Root level parsing
  println("=== Test 5: Root level parsing ===")
  val yamlContent5 = """
simple_key: simple_value
number_key: 999
bool_key: false
"""
  
  val result5 = ViashYamlParser.parseYaml(Some(yamlContent5))
  
  runTest("simple_key", "simple_value", result5.getOrElse("simple_key", None))
  runTest("number_key", 999, result5.getOrElse("number_key", None))
  runTest("bool_key", false, result5.getOrElse("bool_key", None))
  println()

  // Test 6: Edge cases
  println("=== Test 6: Edge cases ===")
  val yamlContent6 = """
par:
  empty_string: ""
  zero: 0
  empty_array:
meta:
  empty_section:
"""
  
  val result6 = ViashYamlParser.parseYaml(Some(yamlContent6))
  val par6 = result6.get("par").map(_.asInstanceOf[Map[String, Any]]).getOrElse(Map.empty)
  val meta6 = result6.get("meta").map(_.asInstanceOf[Map[String, Any]]).getOrElse(Map.empty)
  
  runTest("par.empty_string", "", par6.getOrElse("empty_string", None))
  runTest("par.zero", 0, par6.getOrElse("zero", None))
  runTest("par.empty_array", None, par6.getOrElse("empty_array", None))
  runTest("meta.empty_section", None, meta6.getOrElse("empty_section", None))
  println()

  // Print results
  println("=" * 50)
  println(s"Tests completed: ${testsPass + testsFail}")
  println(s"Tests passed: $testsPass")
  if (testsFail > 0) {
    println(s"Tests failed: $testsFail")
  } else {
    println("All tests passed!")
  }
  
  if (testsFail > 0) {
    sys.exit(1)
  }
}

File: src/test/resources/io/viash/helpers/languages/javascript/test_ViashParseYaml.js
#!/usr/bin/env node

/**
 * Unit test for viashParseYaml function
 * This test verifies that the JavaScript YAML parser correctly parses YAML content
 * and returns appropriate JavaScript data structures.
 */

const fs = require('fs');
const path = require('path');

// Load the YAML parser
const yamlParserPath = path.join(__dirname, '../../../../main/resources/io/viash/languages/javascript/ViashParseYaml.js');
const yamlParserCode = fs.readFileSync(yamlParserPath, 'utf8');

// Since the parser doesn't export when run as main, we need to evaluate it and extract the functions
eval(yamlParserCode);

// Colors for test output
const Colors = {
    RED: '\033[0;31m',
    GREEN: '\033[0;32m',
    YELLOW: '\033[1;33m',
    NC: '\033[0m'  // No Color
};

// Test counters
let testsP = 0;
let testsF = 0;

function runTest(testName, expected, actual) {
    process.stdout.write(`Testing ${testName}... `);
    
    if (JSON.stringify(actual) === JSON.stringify(expected)) {
        console.log(`${Colors.GREEN}PASS${Colors.NC}`);
        testsP++;
    } else {
        console.log(`${Colors.RED}FAIL${Colors.NC}`);
        console.log(`  Expected: ${JSON.stringify(expected)}`);
        console.log(`  Actual:   ${JSON.stringify(actual)}`);
        testsF++;
    }
}

function testArray(testName, expectedArray, actualArray) {
    process.stdout.write(`Testing ${testName}... `);
    
    if (JSON.stringify(actualArray) === JSON.stringify(expectedArray)) {
        console.log(`${Colors.GREEN}PASS${Colors.NC}`);
        testsP++;
    } else {
        console.log(`${Colors.RED}FAIL${Colors.NC}`);
        console.log(`  Expected array: ${JSON.stringify(expectedArray)}`);
        console.log(`  Actual array:   ${JSON.stringify(actualArray)}`);
        testsF++;
    }
}

function main() {
    console.log(`${Colors.YELLOW}Running viashParseYaml JavaScript unit tests...${Colors.NC}`);
    console.log('');

    // Test 1: Basic key-value pairs in sections
    console.log('=== Test 1: Basic key-value pairs ===');
    const yamlContent1 = `
par:
  input: "/path/to/input.txt"
  number: 42
  flag: true
  empty_value: null
meta:
  name: "test_component"
  version: "1.0.0"
`;

    const result1 = viashParseYaml(yamlContent1);
    
    // Test the results
    runTest('par.input', '/path/to/input.txt', result1.par?.input);
    runTest('par.number', 42, result1.par?.number);
    runTest('par.flag', true, result1.par?.flag);
    runTest('par.empty_value', null, result1.par?.empty_value);
    runTest('meta.name', 'test_component', result1.meta?.name);
    runTest('meta.version', '1.0.0', result1.meta?.version);
    console.log('');

    // Test 2: Arrays in sections
    console.log('=== Test 2: Arrays ===');
    const yamlContent2 = `
par:
  files:
    - "file1.txt"
    - "file2.txt" 
    - "file3.txt"
  numbers:
    - 1
    - 2
    - 3
`;

    const result2 = viashParseYaml(yamlContent2);
    
    testArray('par.files', ['file1.txt', 'file2.txt', 'file3.txt'], 
               result2.par?.files || []);
    testArray('par.numbers', [1, 2, 3], 
               result2.par?.numbers || []);
    console.log('');

    // Test 3: Quoted strings and special characters  
    console.log('=== Test 3: Quoted strings ===');
    const yamlContent3 = `
par:
  quoted_string: "Hello \\"World\\""
  single_quoted: 'Single quotes'
  with_newline: "Line 1\\nLine 2"
`;

    const result3 = viashParseYaml(yamlContent3);
    
    runTest('par.quoted_string', 'Hello "World"', result3.par?.quoted_string);
    runTest('par.single_quoted', 'Single quotes', result3.par?.single_quoted);
    runTest('par.with_newline', 'Line 1\nLine 2', result3.par?.with_newline);
    console.log('');

    // Test 4: Boolean and numeric types
    console.log('=== Test 4: Boolean and numeric types ===');
    const yamlContent4 = `
par:
  bool_true: true
  bool_false: false
  integer: 123
  negative: -456
  float_val: 3.14
  negative_float: -2.5
`;

    const result4 = viashParseYaml(yamlContent4);
    
    runTest('par.bool_true', true, result4.par?.bool_true);
    runTest('par.bool_false', false, result4.par?.bool_false);
    runTest('par.integer', 123, result4.par?.integer);
    runTest('par.negative', -456, result4.par?.negative);
    runTest('par.float_val', 3.14, result4.par?.float_val);
    runTest('par.negative_float', -2.5, result4.par?.negative_float);
    console.log('');

    // Test 5: Root level parsing (no sections)
    console.log('=== Test 5: Root level parsing ===');
    const yamlContent5 = `
simple_key: simple_value
number_key: 789
bool_key: true
`;

    const result5 = viashParseYaml(yamlContent5);
    
    runTest('simple_key', 'simple_value', result5.simple_key);
    runTest('number_key', 789, result5.number_key);
    runTest('bool_key', true, result5.bool_key);
    console.log('');

    // Test 6: Edge cases
    console.log('=== Test 6: Edge cases ===');
    const yamlContent6 = `
par:
  empty_string: ""
  zero: 0
  empty_array:
meta:
  empty_section:
`;

    const result6 = viashParseYaml(yamlContent6);
    
    runTest('par.empty_string', '', result6.par?.empty_string);
    runTest('par.zero', 0, result6.par?.zero);
    runTest('par.empty_array', [], result6.par?.empty_array || []);
    runTest('meta.empty_section', null, result6.meta?.empty_section);
    console.log('');

    // Print summary
    const totalTests = testsP + testsF;
    console.log('='.repeat(50));
    console.log(`Tests completed: ${totalTests}`);
    console.log(`${Colors.GREEN}Tests passed: ${testsP}${Colors.NC}`);
    if (testsF > 0) {
        console.log(`${Colors.RED}Tests failed: ${testsF}${Colors.NC}`);
        process.exit(1);
    } else {
        console.log(`${Colors.GREEN}All tests passed!${Colors.NC}`);
    }
}

main();
